keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,854 @@
+are insufficient for many use cases, and aren't easily appended with provider rules.
+
+_What about [CEL for Admission Control](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control)?_
+`ValidatingAdmissionPolicy` is an exciting new feature which we hope will greatly reduce the need
+for admission webhooks, but it is intentionally not attempting to cover every possible use case.
+This proposal aims to improve the situation for those webhooks that cannot be migrated.
+
+### User Stories
+
+#### Exclude resources from a wildcard rule
+
+> I want to enforce metadata policy through an admission webhook without adding latency & risk to
+> high QPS system requests.
+
+Currently, if a webhook uses wildcard match rules, there is no way to filter out a subset of
+resources or requests from matching the wildcard. If the webhook instead enumerates every resource
+that should match, it must be kept up-to-date with every CRD that's added.
+
+With CEL match conditions, the webhook could specify wildcard match rules, and add match conditions
+to filter out the desired resources:
+
+```yaml
+rules:
+  # Match CREATE & UPDATE on all resources:
+  - operations:
+    - CREATE
+    - UPDATE
+    apiGroups: '*'
+    apiVersions: '*'
+    resources: '*'
+matchConditions:
+  # Exclude leases from the webhook
+  - expression: '!(request.resource.group == "coordination.k8s.io" && resource.resource == "leases")'
+```
+
+#### Exempt system users from security policy
+
+> As a managed cluster provider, I want to prevent user webhooks from intercepting critical system
+> requests.
+
+System _resources_ can currently be exempted through a namespace or label selector, but requests by
+system components against non-system resources cannot be. For example, update pod status requests by
+Kubelets cannot be excluded from user webhooks intercepting all pod requests.
+
+With `matchConditions`, a managed cluster could append system-exclusion rules to each webhook. For example:
+
+```yaml
+matchConditions:
+  # Exclude node requests from the webhook
+  - expression: '!("system:nodes" in request.userInfo.groups)'
+```
+
+Since the expression will be evaluated using a common Kubernetes CEL library, these expressions
+should also get automatic access to the secondary authorization check mechanism described in
+[KEP-3488: CEL for Admission Control](/keps/sig-api-machinery/3488-cel-admission-control#secondary-authz).
+In practice, this means that RBAC bindings can be used to opt-out privileged users from security policy:
+
+_Note: The secondary authz mechanism has not yet been designed, and the example syntax here is just
+to illustrate how it might be used._
+
+```yaml
+matchConditions:
+  # Exclude users with the 'breakglass' permission on the 'security-policy' webhook.
+  - expression: '!authorized(request.userInfo, "breakglass", "validatingwebhookconfigurations.admissionregistration.k8s.io/security-policy")'
```

Comment:
This seems like something either a) the people with breakglass permission or b) the cluster admin might want to add to many webhook configurations. But: there's no facility to make it easy or possible for either of those groups to require this of webhook configurations. Is there some further plan?

Reply:
I would definitely like to explore match conditions that are applied across webhooks. Another use case to add to the 2 you listed is if a managed cluster provider wants to exclude managed resources from webhooks. @soorena776 is also looking into this.

I decided to make this out-of-scope for this KEP, since I think it expands the scope quite a bit, but we should definitely discuss it in the future.

I added a `Future work` section.